Engineering/Mobile/Flow in Kotlin.md

```diff
@@ -0,0 +1,263 @@
+---
+tags: mobile, android, kotlin, flow
+author: Truong Hung Khanh
+date: 2022-09-05
+---
+
+## **What is a Flow in Kotlin?**
+
+- It’s a Kotlin language feature that serves as a reactive programming framework
+- It’s all about being notified about changes in your code and sending them through a pipeline that potentially modifies them
+- A Flow is a coroutine that can emit multiple values over a period of time
+
+## **Count-down example**
+
+First, let's see an example of how can we declare a flow. Remember it, we are going to use this example a few times below.
+
+```kotlin
+class MainViewModel: ViewModel() {
+
+	// Create an integer flow
+	val countDownFlow = flow<Int> {
+		val startingValue = 10
+		val currentValue = startingValue
+
+		while(currentValue > 0) {
+			// Because we are currently on a coroutine, we can call delay here
+			delay(1000L)
```

Comment:
Reactive programming usually has "debouce" operator, please research to see if is there any similar operator, using "delay" or "sleep" is bad because it holds the thread. With debouce OS can switch the thread to do another task.